### ISSUE TITLE
Turn on Blake3 for gateways

### ISSUE THREAD

**Issue by elek** [storj/gateway-mt#310]
We implemented Blake3 implementation of the piece hash algorithm (current one is SHA-256)

Blake3 has similar security, but it uses 16% less CPU based on the tests. The wall-clock time of upload is roughly the same. (We couldn't identify statistically significant difference.

We should create a feature flag that can be used to make blake3 the default hash algorithm in gateways

**Comment by amwolff** [comment]
> Blake3 has similar security

What are the differences?

> We should make blake3 as the default hash algorithm in gateways, and test it and deploy it.

How to make it the default?

Overall, it sounds like a good idea!

**Comment by elek** [comment]
> What are the differences?

Let's say, no differences ;-)  What I intended to say: both are cryptographically safe.  (xxh3 even more faster, but not cryptographically safe)

> How to make it the default?

There are multiple options:

* we can make it default in libuplink --> no change, but risk to enable it everywhere
* for enabling it only for special gateways / request, the hash type should be signaled in the context:

https://github.com/storj/uplink/blob/0618502c1396e98afe566b27f34cdc14a4367cce/private/piecestore/hash.go#L15

The challenge here (IMHO) is finding a good (and safe) testing / deployment plan. Eg.: turn it on only for one project or one special satellite first.

**Comment by elek** [comment]
Had an offline conversation with @littleskunk 

Activation plan: 

Today we can enable blake3 with a context variable. (you can download/upload blake3 today with setting an environment variables for uplink cli --> which sets the context variable).

We have the following options:
 * change the default from SHA-256 to Blake3
 * change the default but only for gateway-mt processes
 * introduce a configuration for gateway-mt and change it only for selected gateaway-mt
I think the 3rd one is the reduced scope. (change it only for one gateway-mt first...)


Test plan:

1. Repair test:

 > The minimum test here would be to upload some data, delete files from one node, corrupt files from another node, verify that the audit job detects both problems, verify that the repair job will get it back to full health.

2. Graceful exit

> We should also test if graceful exit transfers the piece to a new storage node without any issues.

3. Backward compatibility test

> We do have a backwards compatibility test that we could just extend a bit. All we need to do is enable blake3 on one of the file uploads. The rest of the test should work out just fine.

4. Unit test

> Do we also have similar unit test for repair and graceful exit?

TODO: check what is covered with unit test

**Comment by ferristocrat** [comment]
I think we should include a performance test before/after. If the goal for gateway is reduced CPU consumption in the same time we should ensure that's what we get:

**Comment by elek** [comment]
We already had multiple performance test, including long term retrievability checker instances which collected historical data from upload/download. This is why we would like to deploy it. Also multiple microbenchmarks.

What is your recommendation @ferristocrat to be measured?

**Comment by ferristocrat** [comment]
The two metrics that edge team would want to at least monitor are CPU load (to validate that the change had the desired effect of reducing this number) and clock time (to ensure we're not degrading customer experience).

**Comment by ferristocrat** [comment]
Related to the Activation/Test Plan above - it looks like we need to wait for testing of other subsystems to be complete before we'd be ready to toggle to Blake3 on Edge production. Is this accurate?

**Comment by amwolff** [comment]
I think we shouldn't bring this to sprint 10, given that it's unclear whether the impact on the network has been fully tested yet and if we can bring it to the gateway.

**Comment by amwolff** [comment]
@ferristocrat @NiaStorj I think we shouldn't bring this to sprint 10, given that it's unclear whether the impact on the network has been fully tested yet and if we can bring it to the gateway.

**Comment by elek** [comment]
We need a patch to make it configurable in gateway-mt (it's a small change to add a value to the context). That modification can be planned to Sprint 10. 

Deployment depends on the status of QA.

**Comment by elek** [comment]
We need this one line to change the used hash algorithm for uploads. The safest option is to make it configurable. I am not sure what is the right place for this in gateway-mt.

```
piecestore.WithPieceHashAlgo(ctx, pb.PieceHashAlgorithm(value))
```

**Comment by iglesiasbrandon** [comment]
@amwolff @ferristocrat @NiaStorj ^ can you create a ticket for the edge team to make this configurable and pull it into sprint 10 if possible?

**Comment by amwolff** [comment]
okay, so the only addition to the acceptance criteria is that turning this on for uploads is feature flagged

**Comment by amwolff** [comment]
@iglesiasbrandon can this issue be doing that? This is gateway-mt's repository, so if we want to represent work needed to test the impact of Blake3's on the network, I think it should be in storj/storj or elsewhere.

**Comment by elek** [comment]
Hi, do you need any help to make this configurable on gateway-mt? That's only setting the flag in the context as I wrote before, not sure about the right place, but let me know if you need help or it's blocked on sg.

**Comment by shaupt131** [comment]
@amwolff @ferristocrat @NiaStorj I modified the ticket with "We should create a feature flag that can be used to make blake3 the default hash algorithm in gateways" so that this ticket now ONLY includes adding the feature flag for Blake3.  We will handle testing/deployment in another ticket.   Will it now be possible to schedule this for sprint 12?

**Comment by shaupt131** [comment]
Since @ferristocrat made https://github.com/storj/gateway-mt/issues/326 and Artur proposes that testing and deployment reside in another repo, can we close this ticket?

**Comment by elek** [comment]
> testing and deployment reside in another repo, can we close this ticket?

What is the _other_ repo to be used for the issue of testing and deployment?

**Comment by shaupt131** [comment]
@amwolff [proposes storj/storj](https://github.com/storj/gateway-mt/issues/310#issuecomment-1468272740), but I guess I'll look to you for the proper place for these things, then I can whip up some tickets.

**Comment by elek** [comment]
> `proposes storj/storj` 

I see, thanks. 

Personally, I feel it closer to `storj/gateway-mt` than `storj/storj`, even if it's a test/deployment task and not a code implementation. Especially if the edge team is the owner of the deployment.

But it's not a very strong opinion, I can follow it anywhere ;-) 

If I understood well, issues also can be moved: 

https://docs.github.com/en/issues/tracking-your-work-with-issues/transferring-an-issue-to-another-repository

**Comment by ferristocrat** [comment]
Charts: https://thanos.storj.rodeo/d/FFhSvSf4k/vital-signs

**Comment by amwolff** [comment]
This is blocked until we get a green light from @storj/qa / @shaupt131.

**Comment by iglesiasbrandon** [comment]
We discussed turning blake3 next week in a point release in our performance office hours yesterday. This will give us a week to gather data on the changes made this week in production. 

@shaupt131 @amwolff @elek
